demos/common/monitors/presenter.cpp

```diff
@@ -0,0 +1,319 @@
+                break;
+            default: throw std::runtime_error("Unknown monitor type");
+        }
+    }
+    return enabledMonitors;
+}
+}
+
+void meansToOstream(const std::map<MonitorType, std::vector<double>> means, std::ostream& stream) {
+    std::ostringstream tmpStream; // create tmp stream to avoid affecting provided stream`s settings
+    const auto cpuAverageMean = means.find(MonitorType::CpuAverage);
+    if (means.end() != cpuAverageMean) {
+        if (cpuAverageMean->second.empty()) {
+            tmpStream << "No data collected for CPU utilization\n";
+        } else {
+            assert (1 == cpuAverageMean->second.size());
+            tmpStream << "Mean CPU utilization: " << std::fixed << std::setprecision(1)
+                << cpuAverageMean->second.front() * 100 << "%\n";
+        }
+    }
+    const auto distributionCpuMean = means.find(MonitorType::DistributionCpu);
+    if (means.end() != distributionCpuMean) {
+        if (distributionCpuMean->second.empty()) {
+            tmpStream << "No data collected per core utilization\n";
+        } else {
+            tmpStream << "Mean core utilization: " << std::fixed << std::setprecision(1);
+            for (double mean : distributionCpuMean->second) {
+                tmpStream << mean * 100 << "% ";
+            }
+            tmpStream << '\n';
+        }
+    }
+    const auto memoryMean = means.find(MonitorType::Memory);
+        if (means.end() != memoryMean) {
+            if (memoryMean->second.empty()) {
+                tmpStream << "No data collected for memory ans swap usage\n";
+            } else {
+                assert (2 == memoryMean->second.size());
+                tmpStream << "Mean memory usage: " << std::fixed << std::setprecision(1) << memoryMean->second.front()
+                    << "GiB\n";
+                tmpStream << "Mean swap usage: " << std::fixed << std::setprecision(1) << memoryMean->second.back()
+                    << "GiB\n";
+            }
+        }
+    stream << tmpStream.str();
+}
+
+Presenter::Presenter(std::set<MonitorType> enabledMonitors,
+        int yPos,
+        cv::Size graphSize,
+        std::size_t historySize) :
+            yPos{yPos},
+            graphSize{graphSize},
+            graphPadding{std::max(1, static_cast<int>(graphSize.width * 0.05))},
+            historySize{historySize},
+            distributionCpuEnabled{false},
+            strStream{std::ostringstream(std::ios_base::app)} {
+    for (MonitorType monitor : enabledMonitors) {
+        addRemoveMonitor(monitor);
+    }
+}
+
+Presenter::Presenter(const std::string& keys, int yPos, cv::Size graphSize, std::size_t historySize) :
+    Presenter{strKeysToMonitorSet(keys), yPos, graphSize, historySize} {}
+
+void Presenter::addRemoveMonitor(MonitorType monitor) {
+    int sampleStep = std::max(1, static_cast<int>(graphSize.width / historySize));
+    unsigned updatedHistorySize = (graphSize.width + sampleStep - 1) / sampleStep; // round up
```

Comment:
Could it be possible for `updatedHistorySize` to be set to 1 here? I think that would break the update logic below.

Reply:
updated